packages/libsql/src/lib/LibSqlClient.ts

```diff
@@ -0,0 +1,203 @@
+import { randomUUID } from "crypto"
+import type { Client, Transaction } from "@libsql/client"
+import { IEventData, ITimelineEvent } from "@equinox-js/core"
+
+type LibSqlWriteResult =
+  | { type: "Written"; position: bigint; snapshot_etag?: string }
+  | { type: "ConflictUnknown" }
+export type Format = string
+
+type Trx = {
+  execute: Client['execute']
+}
+
+export class LibSqlWriter {
+  constructor(private readonly client: Client) {}
+
+  /**
+   * Reads the current version of the stream
+   * version is 0 based, an empty stream has version 0
+   */
+  async readStreamVersion(streamName: string, trx: Trx = this.client): Promise<bigint> {
+    const result = await trx.execute({
+      sql: "select max(position) from messages where stream_name = ?",
+      args: [streamName],
+    })
+    if (result.rows.length === 0) return 0n
+    const position = result.rows[0][0] as string
+    if (position == null) return 0n
+    return BigInt(position)
+  }
+
+  async writeMessages(
+    category: string,
+    streamName: string,
+    messages: IEventData<Format>[],
+    expectedVersion: bigint | null,
+    updateSnapshot?: (trx: Transaction) => Promise<void>,
+  ): Promise<LibSqlWriteResult> {
+    const trx = await this.client.transaction("write")
+    let position = await this.readStreamVersion(streamName, trx)
+    if (expectedVersion != null && position !== expectedVersion) {
+      await trx.rollback()
+      return { type: "ConflictUnknown" }
+    }
+
+    try {
+      for (let i = 0; i < messages.length; ++i) {
+        const message = messages[i]
+        await trx.execute({
+          sql: `insert into messages(id, stream_name, category, type, data, metadata, position) 
+                values (?, ?, ?, ?, ?, ?, ?)`,
+          args: [
+            message.id || randomUUID(),
```

Comment:
is there not an earlier gate that can do this defaulting?
also why does schema not include a timestamp on the event

Reply:
the type of `IEventData` has `id` as an optional field, so this is the only place to do it really.

Yeah, timestamp is generated server side (avoids clock drift on many occasions)

Reply:
could the ctor not fill it in?